src/lib/pipeline/DispatchIssueUnit.cc

```diff
@@ -53,10 +54,66 @@ DispatchIssueUnit::DispatchIssueUnit(
 
 void DispatchIssueUnit::tick() {
   input_.stall(false);
+  ticks_++;
 
   // Reset the counters
   std::fill(dispatches.begin(), dispatches.end(), 0);
 
+  // Check if waiting instructions are ready.
+  size_t index = 0;
+  auto begin = waitingInstructions_.begin();
+  while (index < waitingInstructions_.size()) {
+    // Check if item at index has waited long enough
+    if (std::get<0>(waitingInstructions_[index]) == ticks_) {
+      auto& entry = std::get<0>(std::get<1>(waitingInstructions_[index]));
+      auto value = std::get<1>(std::get<1>(waitingInstructions_[index]));
+
+      entry.uop->supplyOperand(entry.operandIndex, value);
+      if (entry.uop->canExecute()) {
+        // Add the now-ready instruction to the relevant ready queue
+        auto rsInfo = portMapping_[entry.port];
+        reservationStations_[rsInfo.first].ports[rsInfo.second].ready.push_back(
+            std::move(entry.uop));
+      }
+
+      // Remove entry from vector
+      waitingInstructions_.erase(begin + index);
+    } else
+      index++;
+  }
+
+  // Check if uop is ready.
+  index = 0;
+  auto start = dependantInstructions_.begin();
+  while (index < dependantInstructions_.size()) {
+    auto& entry = dependantInstructions_[index];
+    auto& sourceRegisters = entry.uop->getOperandRegisters();
+    const auto& reg = sourceRegisters[entry.operandIndex];
+
+    bool supplied = false;
+
+    if (scoreboard_[reg.type][reg.tag]) {
+      // The scoreboard says it's ready; read and supply the register value
+      entry.uop->supplyOperand(entry.operandIndex, registerFileSet_.get(reg));
+      supplied = true;
+    }
+
+    if (entry.uop->canExecute()) {
+      // Add the now-ready instruction to the relevant ready queue
+      auto rsInfo = portMapping_[entry.port];
+      reservationStations_[rsInfo.first].ports[rsInfo.second].ready.push_back(
+          std::move(entry.uop));
+    }
+
+    // Remove completed instructions from dependantInstructions_ or increase
+    // index
+    if (supplied)
+      dependantInstructions_.erase(start + index);
+    else
+      index++;
```

Comment:
Why aren't we just incrementing the iterator here? Why are we using a seperate index value?